workflows/cloudify_system_workflows/blueprint.py

```diff
@@ -0,0 +1,225 @@
+from cloudify.exceptions import InvalidBlueprintImport, WorkflowFailed
+from cloudify.constants import (CONVENTION_APPLICATION_BLUEPRINT_FILE,
+                                SUPPORTED_ARCHIVE_TYPES)
+
+from dsl_parser import constants, tasks
+from dsl_parser import utils as dsl_parser_utils
+
+
+@workflow
+def upload(ctx, **kwargs):
+
+    client = get_rest_client()
+
+    # extract the execution parameters
+    blueprint_id = kwargs['blueprint_id']
+    app_file_name = kwargs['app_file_name']
+    url = kwargs['url']
+    file_server_root = kwargs['file_server_root']
+
+    # Download the archive, one way or the other
+    archive_target_path = tempfile.mkdtemp()
+    if url:
+        # download the blueprint archive from URL using requests:
+        client.blueprints.update(
+            blueprint_id,
+            update_dict={'state': BlueprintUploadState.UPLOADING})
+        try:
+            with requests.get(url, stream=True, timeout=(5, None)) as resp:
+                resp.raise_for_status()
+                archive_file_path = os.path.join(archive_target_path,
+                                                 resp.url.split('/')[-1])
+                with open(archive_file_path, 'wb') as f:
+                    for chunk in resp.iter_content(chunk_size=8192):
+                        if chunk:
+                            f.write(chunk)
+
+        except requests.exceptions.RequestException as e:
+            client.blueprints.update(
+                blueprint_id,
+                update_dict={'state': BlueprintUploadState.FAILED_UPLOADING,
+                             'error': str(e),
+                             'error_traceback': traceback.format_exc()})
+            remove(archive_target_path)
+            raise e
+
+        # Upload the downloaded archive to the manager
+        client.blueprints.upload_archive(
+            blueprint_id,
+            archive_path=archive_file_path)
+
+    else:
+        # download the blueprint archive using REST
+        archive_file_path = client.blueprints.download(
+            blueprint_id, output_file=archive_target_path)
+
+    ctx.logger.info('Blueprint archive uploaded. Extracting...')
+
+    # Extract the archive so we can parse it
+    client.blueprints.update(
+        blueprint_id, update_dict={'state': BlueprintUploadState.EXTRACTING})
+    try:
+        archive_util.unpack_archive(archive_file_path, archive_target_path)
+    except archive_util.UnrecognizedFormat:
+        error_msg = 'Blueprint archive is of an unrecognized format. ' \
+                    'Supported formats are: ' \
+                    '{0}'.format(SUPPORTED_ARCHIVE_TYPES)
+        handle_failed_extracting(ctx, client, blueprint_id, error_msg,
+                                 archive_target_path)
+    except Exception as e:
+        handle_failed_extracting(ctx, client, blueprint_id, str(e),
+                                 archive_target_path)
+    archive_file_list = os.listdir(archive_target_path)
```

Comment:
What is in this list? Can you maybe assign to variables its contents? `archive_file_list[0]` means nothing to me.

Reply:
done